### ISSUE TITLE
ArmCord 3.2.4: Window turning blank after a couple of seconds

### ISSUE THREAD

**Issue by KaleiAlma** [smartfrigde/armcord#454]
**Describe the bug**
Just as the title states.

**To Reproduce**
1. Open ArmCord
2. Wait for it to launch normally
3. It breaks

**Expected behavior**
A functioning window.

**Screenshots**
I don't think this would help here...

**Desktop (please complete the following information):**
 - OS: Linux (VanillaOS/x86-64)
 - Method of installation: Flatpak (unofficial, I know, keep reading)
 - Version 3.2.4
 - Desktop: X11/GNOME
 - GPU: AMD Ryzen iGPU/NVIDIA GTX 1650 (mobile)

**Additional context**
I just switched from WebCord because of the same exact issue (SpacingBat3/WebCord#452), and came to enjoy the freedom Discord modding brings. This linked issue establishes that it is a problem with Electron 26.

Then it happens on ArmCord.

Aaaand then I check the release notes.
![image](https://github.com/ArmCord/ArmCord/assets/143332769/37999831-6bc5-4c0e-9176-5237e29bb84d)

F

As an addendum: I tried to fix my situation with some of the fixes presented in the linked issue (and also in SpacingBat3/WebCord#444), such as deleting GPU cache (and just attempting to fiddle with GPU settings in the first place), no dice. I also saw something about Pipewire, which is also part of my setup.

Hope this helps.

**Comment by smartfrigde** [comment]
Can you try running with `--disable-gpu` flag?

**Comment by KaleiAlma** [comment]
That didn't do anything to my awareness, though I could just be passing arguments to the flatpak app wrong...

**Comment by KaleiAlma** [comment]
Nvm I think I was doing it right (if `flatpak run xyz.armcord.ArmCord --disable-gpu` is correct form for this)

**Comment by LeSnake04** [comment]
Can confirm on arch linux:

Note I got this, despite armcord being up to date: 
![Screenshot_20230827_164949](https://github.com/ArmCord/ArmCord/assets/47243761/75c3671f-d809-4675-b62c-7fc9b30581d2)

Error message: 
```
(node:3) UnhandledPromiseRejectionWarning: TypeError: Cannot read properties of undefined (reading 'forEach')
    at registerGlobalKeybinds (/app/bin/armcord/resources/app.asar/ts-out/utils.js:418:14)
(Use `armcord --trace-warnings ...` to show where the warning was created)
(node:3) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

...

Error sending from webFrameMain:  Error: Render frame was disposed before WebFrameMain could be accessed                                                                                                      
    at WebFrameMain._sendInternal (node:electron/js2c/browser_init:2:94573)                                                                                                                                   
    at WebContents._sendInternal (node:electron/js2c/browser_init:2:79800)                                                                                                                                    
    at node:electron/js2c/browser_init:2:106942                                                                                                                                                               
    at new Promise (<anonymous>)                                                                                                                                                                              
    at Object.invokeInWebContents (node:electron/js2c/browser_init:2:106689)                                                                                                                                  
    at WebContents.executeJavaScript (node:electron/js2c/browser_init:2:80668)
```

**Comment by KaleiAlma** [comment]
I got the splash asking me to update as well, though I assumed it might've just been Flatpak weirdness at the time. Good to know.

**Comment by KaleiAlma** [comment]
Oddly enough, downgrading on Flatpak reverts the splash to normal (no longer asking me to update).

**Comment by LeSnake04** [comment]
Downgrading with `flatpak update xyz.armcord.ArmCord --commit=319cd48d58e674aefa683da4d6470683fad980bfd1a08de47915234d979ae01c` fixed the issue for me. (Unlike, @KaleiVerita I get the outdated splash though)

**Comment by Cobradabest** [comment]
I have the same issue. Sometimes it just closes the window, though.

> Downgrading with `flatpak update xyz.armcord.ArmCord --commit=319cd48d58e674aefa683da4d6470683fad980bfd1a08de47915234d979ae01c` fixed the issue for me. (Unlike, @KaleiVerita I get the outdated splash though)
This gave me a 404 error.

**Comment by LeSnake04** [comment]
Funnily enough, I cant recreate it on my fedora laptop.

**Comment by smartfrigde** [comment]
Is everyone here on Flatpak?

**Comment by Kyr4l** [comment]
> Is everyone here on Flatpak?

i use the rpm (x64) version and the same issue happens

**Comment by KaleiAlma** [comment]
I think you guys really should pay attention to WebCord's side of things, since it's very evidently the same exact issue.

> Funnily enough, I cant recreate it on my fedora laptop.

This was a hallmark of the issue on WebCord, that it's insanely setup-dependent, hence the listing of my setup details.

I think this problem mostly has to do with GPU setup (check NVIDIA/AMD setups) and potentially changes when run on X11/Wayland,

Regardless, I've been hearing reports that the issue has since disappeared in the newest Electron nightly... can't help but wonder if this was some sort of QA mishap on their end and they're taking their sweet time to fix. Perhaps waiting this one out might be worth it?

**Comment by smartfrigde** [comment]
It seems to be an [upstream bug](https://github.com/electron/electron/issues/39515). There's nothing we can do really apart from waiting for it to get fixed. When it does I'll push out a new release with updated Electron asap. I'll generate new ArmCord builds with older Electron versions (which are hopefully stable) soon and share them here

**Comment by MysticAx0lotl** [comment]
Same issue on Void Linux, running through the terminal shows the following:
```(node:3) UnhandledPromiseRejectionWarning: TypeError: Object has been destroyed
    at IpcMainImpl.<anonymous> (/app/bin/armcord/resources/app.asar/ts-out/ipc.js:94:33)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
(Use `armcord --trace-warnings ...` to show where the warning was created)
(node:3) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
[0828/104559.951383:ERROR:scoped_ptrace_attach.cc(27)] ptrace: Operation not permitted (1)

```

**Comment by KaleiAlma** [comment]
> It seems to be an [upstream bug](https://github.com/electron/electron/issues/39515). There's nothing we can do really apart from waiting for it to get fixed. When it does I'll push out a new release with updated Electron asap. I'll generate new ArmCord builds with older Electron versions (which are hopefully stable) soon and share them here

As I said, yeah...

I think I was slightly frustrated from a supposed lack of awareness (was thinking "if they saw this happen to WebCord they should've held off on updating"), but I already knew there was little else you could do. Though, I also knew it'd be helpful to open an issue where I'd already virtually figured it out (at least on this level), instead of someone posting an issue like "it doesn't work" and it taking a bit more duplicate time and effort to get to this same conclusion. Didn't really cost me anything to do this myself, anyways.

**Comment by dxcvvxd** [comment]
> Is everyone here on Flatpak?

i use the aur version on artix and this is what happens to me (its just this after splash screen)
![image](https://github.com/ArmCord/ArmCord/assets/77129815/7218e76f-04b3-4c3e-a13e-6faf287a2802)

Edit: this is the error btw
```
Error sending from webFrameMain:  Error: Render frame was disposed before WebFrameMain could be accessed
    at WebFrameMain._sendInternal (node:electron/js2c/browser_init:2:94573)
    at WebContents._sendInternal (node:electron/js2c/browser_init:2:79800)
    at node:electron/js2c/browser_init:2:106942
    at new Promise (<anonymous>)
    at Object.invokeInWebContents (node:electron/js2c/browser_init:2:106689)
    at WebContents.executeJavaScript (node:electron/js2c/browser_init:2:80668)
```

**Comment by tazz4843** [comment]
I'm experiencing a similar issue, where the app splash screen is entirely blank. Running in terminal shows these lines that may be relevant
```
[41469:0829/113948.860767:ERROR:gpu_process_host.cc(957)] GPU process exited unexpectedly: exit_code=139
[41469:0829/113948.860800:WARNING:gpu_process_host.cc(1327)] The GPU process has crashed 1 time(s)
[41469:0829/113950.192565:INFO:CONSOLE(5700)] "%c[ProcessUtilsElectron] 
font-weight: bold;
color: purple;
 setMemoryInformation not available.", source: https://canary.discord.com/assets/7904a4dbb1491125a7b1.js (5700)
[41469:0829/113950.782104:ERROR:gpu_process_host.cc(957)] GPU process exited unexpectedly: exit_code=139
[41469:0829/113950.782131:WARNING:gpu_process_host.cc(1327)] The GPU process has crashed 2 time(s)
  ~~snip~~
 [BUILD INFO] Release Channel: canary, Build Number: 223502, Version Hash: 4b07b8a10391ccebe593ff65b6614bd82afe5a45", source: https://canary.discord.com/assets/7904a4dbb1491125a7b1.js (5700)
[41469:0829/113950.923881:INFO:CONSOLE(2)] "AppCrashedFatalReport: getLastCrash not supported.", source: WebpackModule146577 (2)
[41469:0829/113951.464760:INFO:CONSOLE(61)] "Uncaught TypeError: Cannot read properties of null (reading 'addEventListener')", source: /opt/armcord/resources/app.asar/ts-out/preload/settings.js (61)
[41469:0829/113951.481714:INFO:CONSOLE(7)] "Uncaught (in promise) Error: Cannot find module './notosans-400-normalitalic.woff2'", source: https://canary.discord.com/assets/7904a4dbb1491125a7b1.js (7)
[41469:0829/113952.857964:ERROR:gpu_process_host.cc(957)] GPU process exited unexpectedly: exit_code=139
[41469:0829/113952.857992:WARNING:gpu_process_host.cc(1327)] The GPU process has crashed 3 time(s)
[41715:0829/113952.868254:WARNING:vaapi_wrapper.cc(806)] VAAPI video acceleration not available for disabled
[41469:0829/113952.917419:WARNING:gpu_process_host.cc(980)] Reinitialized the GPU process after a crash. The reported initialization time was 0 ms
Renderer process crashed - see https://www.electronjs.org/docs/tutorial/application-debugging for potential debugging information.
Renderer process crashed - see https://www.electronjs.org/docs/tutorial/application-debugging for potential debugging information.

```
If I Ctrl+C the terminal process, the app continues launching from what appears to be a new forked clone of itself, but shows a blank screen in the end.

Using the AUR `armcord-bin` package. `--disable-gpu` has no effect. Same logs appear either way. Upstream Electron issue shows these errors as well.

**Comment by dxcvvxd** [comment]
@tazz4843 this is weird because i use the same package, though i dont get this error, i only get the one i had sent beforehand.

**Comment by NextWork123** [comment]
> It seems to be an [upstream bug](https://github.com/electron/electron/issues/39515). There's nothing we can do really apart from waiting for it to get fixed. When it does I'll push out a new release with updated Electron asap. I'll generate new ArmCord builds with older Electron versions (which are hopefully stable) soon and share them here

Hello, it can be easily fixed if you package armcord along with the alpha version of electron27 (the issue has already been resolved, as I tested it through one of my friends). As a dependency for the AUR package, you can use electron27-bin.

Best Regards,
NextWorks

**Comment by starchyunderscore** [comment]
same issue using the aur version on arch linux. oddly, downgrading to 3.2.3 does not fix it. neither does the suggestion of using --disable-gpu (though that might be flatpak specific?) the errors seem to differ slightly

```
[12515:0829/164720.810885:ERROR:quota_database.cc(964)] Could not open the quota database, resetting.
[12549:0829/164724.321971:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 2 times!
Error sending from webFrameMain:  Error: Render frame was disposed before WebFrameMain could be accessed
    at WebFrameMain._sendInternal (node:electron/js2c/browser_init:2:94573)
    at WebContents._sendInternal (node:electron/js2c/browser_init:2:79800)
    at node:electron/js2c/browser_init:2:106942
    at new Promise (<anonymous>)
    at Object.invokeInWebContents (node:electron/js2c/browser_init:2:106689)
    at WebContents.executeJavaScript (node:electron/js2c/browser_init:2:80668)
[12549:0829/164730.257238:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 3 times!
```
for running without --disable-gpu

```
[13008:0829/164930.237480:ERROR:quota_database.cc(964)] Could not open the quota database, resetting.
Loaded AMOLED-Cord made by LuckFire#4800
[Config manager] tray: false
[Config manager] dynamicIcon: false
Error sending from webFrameMain:  Error: Render frame was disposed before WebFrameMain could be accessed
    at WebFrameMain._sendInternal (node:electron/js2c/browser_init:2:94573)
    at WebContents._sendInternal (node:electron/js2c/browser_init:2:79800)
    at node:electron/js2c/browser_init:2:106942
    at new Promise (<anonymous>)
    at Object.invokeInWebContents (node:electron/js2c/browser_init:2:106689)
    at WebContents.executeJavaScript (node:electron/js2c/browser_init:2:80668)
```
for running with --disable-gpu
```
[13659:0829/165503.049284:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 2 times!
[13625:0829/165503.762809:ERROR:quota_database.cc(964)] Could not open the quota database, resetting.
Loaded AMOLED-Cord made by LuckFire#4800
[Config manager] tray: false
[Config manager] dynamicIcon: false
[13659:0829/165504.267272:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 3 times!
```
for running the downgraded version

**Comment by dxcvvxd** [comment]
@NextWork123 can you describe the process of installation of armcord that way? i'm very interested in how'd you pull that off.

**Comment by NextWork123** [comment]
Hello @dxcvvxd, sorry for the delay.

Anyway, you need to install the `electron27-bin` package from the AUR. After that, edit the `/usr/bin/armcord` file. You will see the path of Electron that you need to put as follows:  `electron = /usr/lib/electron27/electron`. Then, try to start armcord. :)

**Comment by starchyunderscore** [comment]
@NextWork123, I cannot seem to find the path to electron that I would need to edit. I have tried searching for `electron=` `electron =` `/electron` `electron/` and `/usr/lib`, none of which have any results in `/usr/bin/armcord`. I have also tried looking through `~/.configl/ArmCord`, but no luck in any of the files there either. Is it just a case of putting it wherever, or is there a specific spot that you could point me to?

**Comment by NextWork123** [comment]
> @NextWork123, I cannot seem to find the path to electron that I would need to edit. I have tried searching for `electron=` `electron =` `/electron` `electron/` and `/usr/lib`, none of which have any results in `/usr/bin/armcord`. I have also tried looking through `~/.configl/ArmCord`, but no luck in any of the files there either. Is it just a case of putting it wherever, or is there a specific spot that you could point me to?

You are correct. I forgot to mention that it requires `armcord-git` from AUR, as `armcord-bin` has its electron bundled.

**Comment by nyabinary** [comment]
Yep just got an update on NixOS, and it's borked, lol.

**Comment by tazz4843** [comment]
Switching to Electron 27 as suggested above does partly fix it. ArmCord's splash screen is blank for about three seconds, then it renders normally, and the app itself launches normally. Logs have the same "GPU process exited unexpectedly" as before, but it's mostly fixed after that initial issue.

**Comment by PickMeNow** [comment]
> > Is everyone here on Flatpak?
> 
> i use the aur version on artix and this is what happens to me (its just this after splash screen) ![image](https://user-images.githubusercontent.com/77129815/264014320-7218e76f-04b3-4c3e-a13e-6faf287a2802.png)
> 
> Edit: this is the error btw
> 
> ```
> Error sending from webFrameMain:  Error: Render frame was disposed before WebFrameMain could be accessed
>     at WebFrameMain._sendInternal (node:electron/js2c/browser_init:2:94573)
>     at WebContents._sendInternal (node:electron/js2c/browser_init:2:79800)
>     at node:electron/js2c/browser_init:2:106942
>     at new Promise (<anonymous>)
>     at Object.invokeInWebContents (node:electron/js2c/browser_init:2:106689)
>     at WebContents.executeJavaScript (node:electron/js2c/browser_init:2:80668)
> ```

Experiencing the same issue on ArchLinux using ArmCord-3.2.4.AppImage, went back to 3.2.3 no issues

**Comment by Dafz1010** [comment]
Got the same issue here on Arch

**Comment by LeSnake04** [comment]
> Funnily enough, I cant recreate it on my fedora laptop.

I can confirm the issue with the flatpak on KDE x11, gnome Wayland and KDE Wayland on my EndeavourOS system. My laptop with fedora 38 using gnome Wayland doesn't have the issue.

**Comment by zebn0710** [comment]
@NextWork123  I tried it but and i can get in but it flickers and it unusable

**Comment by tazz4843** [comment]
I can reproduce this issue on my laptop (Arch Linux, AMD Ryzen 7 6800U with Radeon iGPU, `amdgpu` driver) but *cannot* reproduce this on my desktop (Arch Linux, 13700K + Intel Arc A770, `i915` driver, using dGPU). If you really want to get this fixed, don't want to patch `electron27` in, and are willing to deal with a performance hit, switch to your Intel CPU's iGPU perhaps?

**Comment by starchyunderscore** [comment]
> switch to your Intel CPU's iGPU perhaps?

someone who only has an intel igpu here: doesnt work. might be becuase its the new one (iris xe) not the older one (intel uhd)?

**Comment by MysticAx0lotl** [comment]
> someone who only has an intel igpu here: doesnt work. might be becuase its the new one (iris xe) not the older one (intel uhd)?

This is what I have as well. I assume we're both on laptops?

**Comment by starchyunderscore** [comment]
> This is what I have as well. I assume we're both on laptops?

yep, lenovo yoga 7i. 11th gen intel i5, arch linux.

**Comment by zebn0710** [comment]
so its not a Nvidia GPU issue  so hmm

**Comment by zebn0710** [comment]
any idea what can fix this issue  on arch linux

**Comment by tazz4843** [comment]
> someone who only has an intel igpu here: doesnt work. might be becuase its the new one (iris xe) not the older one (intel uhd)?

Interesting. I'm using my A770 not my 13700K but Alchemist is a super new architecture and I would expect it to exhibit the issue as well then.

**Comment by tr1p0p** [comment]
Yeah. Same bug, cannot just setup the client on fedora 38 with X11 or wayland, with a nvidia GPU.


EDIT : Do `flatpak update xyz.armcord.ArmCord --commit=319cd48d58e674aefa683da4d6470683fad980bfd1a08de47915234d979ae01c` fixed the issue for me (well, it's just a revert, but yeah)

Too bad, another broken discord client on linux to add on the list...

**Comment by smartfrigde** [comment]
Please use 3.2.3 for now. There's nothing we can do apart from waiting for Electron to fix or we can downgrade to last Electron version with no issues however that version also proved to have issues on other platforms

**Comment by smartfrigde** [comment]
ArmCord 3.2.4 with Electron 27 Beta 2 (which should fix the issue) is gonna be released soon with pre-release tag on this repo (no build servers this time, like before). Sorry for the wait I was really busy irl and generally thought Electron is gonna fix this upstream on E26 which probably won't happen. I'm gonna link the release once it's done

**Comment by smartfrigde** [comment]
https://github.com/ArmCord/ArmCord/releases/tag/v3.2.4-e27
Should be fixed, enjoy.

**Comment by MysticAx0lotl** [comment]
> https://github.com/ArmCord/ArmCord/releases/tag/v3.2.4-e27 Should be fixed, enjoy.

Has the Flatpak version not been updated? I'm still getting this issue...

**Comment by orowith2os** [comment]
The Flatpak will update when a stable release is out. That is a prerelease. If you feel it's necessary, you can easily modify the Flatpak to pull the Electron 27 builds temporarily.

**Comment by zebn0710** [comment]
use armcord-git from the aur if you  use arch it works when  you install the electron 17-bin

**Comment by MysticAx0lotl** [comment]
> use armcord-git from the aur if you use arch it works when you install the electron 17-bin

I use Void

> If you feel it's necessary, you can easily modify the Flatpak to pull the Electron 27 builds temporarily.

How would I go about doing this? Thanks!

**Comment by orowith2os** [comment]
> How would I go about doing this? Thanks!

Go to https://github.com/flathub/xyz.armcord.armcord and clone it. Look at `xyz.armcord.ArmCord.yml` and modify the link and checksum of the ArmCord binaries to point to the Electron 27 ones (you just need to change the `x86_64` ones). Then build with `flatpak run --filesystem=$(pwd) org.flatpak.Builder --user --force-clean build-dir --install xyz.armcord.ArmCord.yml`.

**Comment by MysticAx0lotl** [comment]
Sweet, ty!

**Comment by dxcvvxd** [comment]
well, it looks like 3.2.4-e27 is the only thing we'll get, because downloading the normal bin version from aur isnt gonna help.
sticking to git version then.

**Comment by smartfrigde** [comment]
I'm gonna push a new stable release (the one which repos will also use) use once E27 is stable. Last time we used a beta Electron release for stable ArmCord release, it ended up in bricked installs.

**Comment by NebulaBC** [comment]
Edited this comment because I did some more digging.

I created this comment to point out that I too am having this issue with the latest release. I've now noticed that the latest release goes back to E26 with the security update for libwebp. Is there any chance that we can get an official E27-beta.9 build of ArmCord (Seeing that it's already changed in dev)? From what I can see beta.9 uses version 118.0.5993.32 of chromium which released earlier today. This should mean it's secure against libwebp's exploit as opposed to the pre-existing beta.2 ArmCord build.
